subsystems/DrivetrainBase.java

```diff
@@ -19,33 +26,99 @@
-        super(new PIDController(turnFPID.kP, turnFPID.kI, turnFPID.kD));
-        
-        this.leftLeader = leftLeader;
-        this.rightLeader = rightLeader;
-        this.leftFollower = leftFollower;
-        this.rightFollower = rightFollower;
-        this.turnFPID = turnFPID;
-        this.turnFPIDThrottle = turnFPIDThrottle;
-        this.invertGyro = invertGyro;
-        leftGroup = new SpeedControllerGroup(leftLeader, leftFollower);
-        rightGroup = new SpeedControllerGroup(rightLeader, rightFollower);
+    // MotorIDs: a arrays of integer CAN IDs used to make motors. index 0 should ALWAYS be the leader motor, and anything else is a follower.
+    public DrivetrainBase(DriveConstants constants, int[] leftMotorIDs, int[] rightMotorIDs) {
+        super(new PIDController(constants.turnFPID.kP, constants.turnFPID.kI, constants.turnFPID.kD));        
+        // disable PID control on start
+        this.disable();
+        this.constants = constants;
+        // int i = 0;
+        // make arrays to put each type of motor in
+        WPI_TalonFX[] leaders = new WPI_TalonFX[2];
+        WPI_TalonFX[] leftFollowers = new WPI_TalonFX[leftMotorIDs.length-1];
+        WPI_TalonFX[] rightFollowers = new WPI_TalonFX[rightMotorIDs.length-1];
+        // put each motor in an array corresponding to its type (leader or follower)
+        for(int i = 0; i < leftMotorIDs.length; i++) {
+            // if it's the first one in the array, it's a leader
+            if(i == 0) {
+                // make a new motor with the given ID and put it in the array
+                leaders[0] = new WPI_TalonFX(leftMotorIDs[i]);
+            }
+            // otherwise it's a follower
+            else {
+                leftFollowers[i-1] = new WPI_TalonFX(leftMotorIDs[i]);
+            }
+        }
+        // do the same for the right motors
+        for(int i = 0; i < rightMotorIDs.length; i++) {
+            if(i == 0) {
+                leaders[1] = new WPI_TalonFX(rightMotorIDs[i]);
+            }
+            else {
+                rightFollowers[i-1] = new WPI_TalonFX(rightMotorIDs[i]);
+            }
+        }
+        // the first leader will be the left one
+        this.leftLeader = leaders[0];
+        // the second leader will be the right one
+        this.rightLeader = leaders[1];
+        // set all variables declared at the top to those given in the constructor (mostly constants)
+        var currentAngle = Rotation2d.fromDegrees(getHeading360());
+        this.odometry = new DifferentialDriveOdometry(currentAngle);
+        this.poseHistory =  new LimitedPoseMap(constants.historyLimit);
+        this.turnFPID = constants.turnFPID;
+        this.turnFPIDThrottle = (double) constants.turnPIDThrottle;
+        this.invertGyro = (boolean) constants.invertGyro;
```

Comment:
You shouldn't need to cast here right?

Reply:
Yeah a lot of this is just left over from the hashmap stuff